src/main/java/net/neoforged/neoforge/common/extensions/IAttributeExtension.java

```diff
@@ -0,0 +1,217 @@
+     */
+    @Nullable
+    default ResourceLocation getBaseId() {
+        if (this == Attributes.ATTACK_DAMAGE.value()) return AttributeUtil.BASE_ATTACK_DAMAGE_ID;
+        else if (this == Attributes.ATTACK_SPEED.value()) return AttributeUtil.BASE_ATTACK_SPEED_ID;
+        else if (this == Attributes.ENTITY_INTERACTION_RANGE.value()) return AttributeUtil.BASE_ENTITY_REACH_ID;
+        return null;
+    }
+
+    /**
+     * Converts an attribute modifier into its tooltip representation.
+     * <p>
+     * This method does not handle formatting of "base" modifiers, such as Attack Damage or Attack Speed.
+     * <p>
+     *
+     * @param modif The attribute modifier being converted to a component.
+     * @param flag  The tooltip flag.
+     * @return The component representation of the passed attribute modifier.
+     */
+    default MutableComponent toBaseComponent(double value, double entityBase, boolean merged, TooltipFlag flag) {
+        Attribute attr = self();
+
+        Component debugInfo = CommonComponents.EMPTY;
+
+        if (flag.isAdvanced() && !merged) {
+            // Advanced Tooltips cause us to emit the entity's base value and the base value of the item.
+            debugInfo = Component.literal(" ")
+                    .append(Component.translatable("neoforge.adv.base", FORMAT.format(entityBase), FORMAT.format(value - entityBase)).withStyle(ChatFormatting.GRAY));
+        }
+
+        MutableComponent comp = Component.translatable("attribute.modifier.equals.0", FORMAT.format(value), Component.translatable(attr.getDescriptionId()));
+
+        return comp.append(debugInfo);
+    }
+
+    /**
+     * Returns the color used by merged attribute modifiers. Only used when {@link NeoForgeMod#enableMergedAttributeTooltips()} is active.
+     * <p>
+     * Similarly to {@link Attribute#getStyle(boolean)}, this method should return a color based on the attribute's {@link Sentiment}.
+     * The returned color should be distinguishable from the color used by {@link Attribute#getStyle(boolean)}.
+     * 
+     * @param positive If the attribute modifier value is positive or not.
+     */
+    TextColor getMergedStyle(boolean isPositive);
+
+    /**
+     * Certain attributes, such as Attack Damage, are increased by an Enchantment that doesn't actually apply
+     * an attribute modifier.<br>
+     * This method allows for including certain additional variables in the computation of "base" attribute values.
+     *
+     * @param stack The stack in question.
+     * @return Any bonus value to be applied to the attribute's value, after all modifiers have been applied.
+     * 
+     * @apiNote This method is experimental as it only exists to support displaying Sharpness in item tooltips, which is currently broken by MC-271840
+     */
+    @ApiStatus.Experimental
+    default double getBonusBaseValue(ItemStack stack) {
+        // if (this == Attributes.ATTACK_DAMAGE) return EnchantmentHelper.getDamageBonus(stack, MobType.UNDEFINED);
```

Comment:
Why is this commented out line here? If it is to show an example of what the javadocs are talking about, then could we move it to a code block in the javadocs?

Reply:
This is the 1.20 implementation for this thing. It has no 1.21 equivalent due to MC-271840. I debated removing it, but it took me long enough to understand that I just left the thing here as experimental rather than try to reimplement it entirely if the vanilla bug gets fixed in a similar manner to 1.20's impl.

We could just drop it, IMO it's a hack, and I hope Mojang doesn't resurrect it this way.

Reply:
I think we can just drop it by doing the following:
1. commit to the PR removing it
2. second commit that adds a comment in the place where it would be used, that both references MC-271840 and links to `https://github.com/neoforged/NeoForge/pull/1551/commit/...` (the commit that removed it)

That way if Mojang *does* end up resurrecting this way, we don't have to spend the time figuring it out again, but if they don't then we don't have this dead commented out code that we never used just hanging around.